Partial solutions

rrsadykov opened the pull request:
- Partial solutions are introduced
- Added support of partial solutions in ColumnGeneration, SolveIpForm, SolveLpForm algorithms
- Correction in calculation of column reduced costs
- Corrections in preprocessing
- Correction in storages restoration
- New test with a user defined algorithm (to test partial solutions and preprocessing)

Branch name is an old one, no diving here.

rrsadykov commented:
@guimarqu Can you please see the preprocessing tests on Julia 1.0. It says `type CachingOptimizer has no field inner`
for the lines
```
coluna_optimizer = problem.moi_backend
master = CL.getmaster(coluna_optimizer.inner.re_formulation)

```

rrsadykov commented:
The tests pass for Julia 1.5, but do not pass for Julia 1.0. The tests were done by Teobaldo I suppose, I do not understand them much

guimarqu commented:
Hi Ruslan, I'll look the PR monday


> @guimarqu Can you please see the preprocessing tests on Julia 1.0. It says `type CachingOptimizer has no field inner`
> for the lines
> 
> ```
> coluna_optimizer = problem.moi_backend
> master = CL.getmaster(coluna_optimizer.inner.re_formulation)
> ```

This bug might be related to issue #376. Is it only with Julia 1.0 ?

guimarqu commented:
Tests have passed with julia 1.0 : https://travis-ci.com/github/atoptima/Coluna.jl/builds/189223931

rrsadykov commented:
Yes, it is strange, now tests pass, I have corrected one thing which is not supported in Julia 1.0 in the new test. 

However, the preprocessing tests still do not pass on my computer with Julia 1.0., but they pass with Julia 1.5. I agree with the issue #376 that we need to update these tests. May be it is better to remove these preprocessing tests. Now preprocessing is tested in the new test I have added (user_algorithms_tests).

guimarqu commented:
Ok so let's comment them in the meantime.
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
Le samedi, 10 octobre 2020 at 10:55 PM, Ruslan Sadykov <notifications@github.com> a écrit :

> Yes, it is strange, now tests pass, I have corrected one thing which is not supported in Julia 1.0 in the new test.
>
> However, the preprocessing tests still do not pass on my computer with Julia 1.0., but they pass with Julia 1.5. I agree with the issue [#376](https://github.com/atoptima/Coluna.jl/issues/376) that we need to update these tests. May be it is better to remove these preprocessing tests. Now preprocessing is tested in the new test I have added (user_algorithms_tests).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/atoptima/Coluna.jl/pull/396#issuecomment-706610330), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABNM5B5YWDGIHS37MIYTEELSKDC45ANCNFSM4SLIVKGA).

guimarqu commented:
Do you have the last version of BlockDecomposition in your julia 1.0 env ?

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
Le samedi, 10 octobre 2020 at 10:57 PM, Guillaume Marques <guillaume.marques@protonmail.com> a écrit :

> Ok so let's comment them in the meantime.
> ‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
> Le samedi, 10 octobre 2020 at 10:55 PM, Ruslan Sadykov <notifications@github.com> a écrit :
>
>> Yes, it is strange, now tests pass, I have corrected one thing which is not supported in Julia 1.0 in the new test.
>>
>> However, the preprocessing tests still do not pass on my computer with Julia 1.0., but they pass with Julia 1.5. I agree with the issue [#376](https://github.com/atoptima/Coluna.jl/issues/376) that we need to update these tests. May be it is better to remove these preprocessing tests. Now preprocessing is tested in the new test I have added (user_algorithms_tests).
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, [view it on GitHub](https://github.com/atoptima/Coluna.jl/pull/396#issuecomment-706610330), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABNM5B5YWDGIHS37MIYTEELSKDC45ANCNFSM4SLIVKGA).

rrsadykov commented:
I have updated BlockDecomposition, still the preprocessing tests do not pass on on my machine with Julia 1.0.

There are also warnings:
```
WARNING: could not import MathProg.getfeasibilitystatus into Algorithm
WARNING: could not import MathProg.setfeasibilitystatus! into Algorithm
WARNING: could not import MathProg.isfeasible into Algorithm
```  
They are related to your previous merge

rrsadykov commented:
I have started to look at the preprocessing in more detail. There are a lot of corrections to make. I will submit them in separate pull request (the old preprocessing tests will be deactivated there). I think we can merge this pull request as it is now, as it passes all checks.

guimarqu reviewed (commented):
I won't merge this branch into master. I'll create a protected branch for v0.4 in which we will merge partial solutions and non-robust cuts so we can gradually improve the code without breaking things in 0.3.

guimarqu commented on src/Algorithm/Algorithm.jl:
```diff
@@ -16,8 +16,8 @@ const MOI = MathOptInterface
-import ..MathProg: getterminationstatus,
-    setterminationstatus!, get_ip_primal_bound, get_ip_dual_bound,
+import ..MathProg: getfeasibilitystatus, getterminationstatus, setfeasibilitystatus!,
+    setterminationstatus!, isfeasible, get_ip_primal_bound, get_ip_dual_bound,
```
I think it's related to your comment about the WARNING. These methods have been deleted.

rrsadykov replied:
I do not remember I added this, but never mind, I will delete.

rrsadykov commented on test/Project.toml:
```diff
@@ -8,3 +8,4 @@ MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 LightGraphs = "093fc24a-ae57-5d10-9952-331d41423f4d"
+Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
```
This to make a cleaner code for the user algorithm test (to use `@with_kw`).

guimarqu replied:
Ah sorry, it's in test so it's ok !

rrsadykov commented on src/Algorithm/colgen.jl:
```diff
@@ -296,7 +299,7 @@ function solve_sp_to_gencol!(
                     else
                         msg = """
-                        Column already exists as $(getname(masterform, col_id)) and is already active.
+                            Column already exists as $(getname(masterform, col_id)) and is already active.
```
ok

rrsadykov commented on src/Algorithm/colgen.jl:
```diff
@@ -284,6 +286,7 @@ function solve_sp_to_gencol!(
     sense = getobjsense(masterform)
     if nb_ip_primal_sols(sp_optstate) > 0
         bestsol = get_best_ip_primal_sol(sp_optstate)
+        #@show bestsol compute_red_cost(algo, masterform, spinfo, bestsol, dualsol)
```
ok

rrsadykov commented on src/Algorithm/colgen.jl:
```diff
@@ -254,7 +255,8 @@ function compute_red_cost(
         red_cost = getvalue(spsol)
     end
-    red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
+    #red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
```
I will remove

rrsadykov commented on src/Algorithm/colgen.jl:
```diff
@@ -21,7 +21,7 @@ restricted master and `pricing_prob_solve_alg` to solve the subproblems.
 """
 @with_kw struct ColumnGeneration <: AbstractOptimizationAlgorithm
-    restr_master_solve_alg = SolveLpForm(get_dual_solution = true)
+    restr_master_solve_alg = SolveRestrMasterInColGen()
```
ok

rrsadykov commented on src/Algorithm/basic/solveipform.jl:
```diff
@@ -56,9 +60,17 @@ function run!(algo::SolveIpForm, data::ModelData, input::OptimizationInput)::Opt
+        partsolstorage = getstorage(data, PartialSolutionStoragePair)
+        partial_solution = get_primal_solution(partsolstorage, form)
+    else    
+        partial_solution = EmptyPrimalSolution(form)
```
Why two arrays?
I agree that most of the time, partial solution is empty. But I do not know how to make it lighter. If you have an idea..

guimarqu replied:
This what `EmptyPrimalSolution()` does. We can keep it like that for now. We will try to improve this gradually.

rrsadykov commented on src/Algorithm/basic/solveipform.jl:
```diff
@@ -73,11 +85,15 @@ function run!(algo::SolveIpForm, data::ModelData, input::OptimizationInput)::Opt
     return OptimizationOutput(optstate)
 end
 
+run!(algo::SolveIpForm, data::ReformData, input::OptimizationInput) = 
```
This is to call primal heuristics from the conquer algorithm. Some heuristics may need to work with reformulation. So the restricted master IP heuristic should support the call with `ReformData`.

guimarqu replied:
Ok

rrsadykov commented on src/Algorithm/conquer.jl:
```diff
@@ -186,7 +218,7 @@ end
 
 @with_kw struct RestrMasterLPConquer <: AbstractConquerAlgorithm 
-    masterlpalgo::SolveLpForm = SolveLpForm()
+    masterlpalgo::SolveLpForm = SolveLpFormInRestrMasterLPConquer()
```
ok

rrsadykov commented on src/MathProg/solutions.jl:
```diff
@@ -49,6 +53,24 @@ function contains(sol::DualSolution, f::Function)
     return false
 end
 
+function concatenate_sols(sola::PrimalSolution{M}, solb::PrimalSolution{M}) where {M}
```
yes, why not.

guimarqu replied:
Ok we keep this for future PR.

rrsadykov commented on src/MathProg/formulation.jl:
```diff
@@ -637,3 +637,24 @@ function write_to_LP_file(form::Formulation, filename::String)
     end
 end
+
+function getspsol(master::Formulation{DwMaster}, col_id::VarId)
```
I was using it while debugging. Can we just comment?

guimarqu replied:
Ok

rrsadykov commented:
I will do the corrections tomorrow

rrsadykov commented on src/Algorithm/utilities/optimizationstate.jl:
```diff
@@ -127,7 +127,8 @@ end
 
 function CopyBoundsAndStatusesFromOptState(
-    form::AbstractFormulation, source_state::OptimizationState, copy_ip_primal_sol::Bool
+    form::AbstractFormulation, source_state::OptimizationState, 
```
I have renamed this function to `OptimizationState` I have also commented previous function in this file, as it was not used

rrsadykov commented:
Corrections are made
